core-processor/src/processor.rs

```diff
@@ -42,33 +49,197 @@ enum SuccessfulDispatchResultKind {
     Success,
 }
 
-/// Process program & dispatch for it and return journal for updates.
-pub fn process<A: ProcessorExt + EnvExt + IntoExtInfo + 'static, E: Environment<A>>(
+/// Checked parameters for message execution across processing runs.
+pub struct PreparedMessageExecutionContext {
+    gas_counter: GasCounter,
+    gas_allowance_counter: GasAllowanceCounter,
+    dispatch: IncomingDispatch,
+    origin: ProgramId,
+    balance: u128,
+    program: Program,
+    memory_size: WasmPageNumber,
+}
+
+impl PreparedMessageExecutionContext {
+    /// Returns reference to the GasCounter.
+    pub fn gas_counter(&self) -> &GasCounter {
+        &self.gas_counter
+    }
+}
+
+/// Defines result variants of the function `prepare`.
+#[allow(clippy::large_enum_variant)]
+pub enum PrepareResult {
+    /// Successfully precharged for memory pages.
+    Ok {
+        /// A context for `process` function.
+        context: PreparedMessageExecutionContext,
+        /// A set with numbers of memory pages which contain some data.
+        pages_with_data: BTreeSet<PageNumber>,
+    },
+    /// Required function is not exported. The program will not be executed.
+    WontExecute(Vec<JournalNote>),
+    /// Provided actor is not executable or there is not enough gas for memory pages size.
+    Error(Vec<JournalNote>),
+}
+
+/// Prepares environment for the execution of a program.
+/// Checks if there is a required export and tries to pre-charge for memory pages.
+/// Returns either `PreparedMessageExecutionContext` for `process` or an array of journal notes.
+/// See `PrepareResult` for details.
+pub fn prepare(
     block_config: &BlockConfig,
     execution_context: MessageExecutionContext,
-) -> Vec<JournalNote> {
+) -> PrepareResult {
     let MessageExecutionContext {
         actor,
         dispatch,
         origin,
         gas_allowance,
+        subsequent_execution,
     } = execution_context;
     let Actor {
         balance,
         destination_program,
         executable_data,
     } = actor;
 
-    match check_is_executable(executable_data, &dispatch) {
-        Err(exit_code) => process_non_executable(dispatch, destination_program, exit_code),
-        Ok(data) => process_executable::<A, E>(
-            origin,
-            gas_allowance,
-            data,
+    let data = match check_is_executable(executable_data, &dispatch) {
```

Comment:
can't we destruct data struct into fields here?
like Data {
   program,
   pages_with_data
}

etc